### PR TITLE
Mainnet fork

### PR DESCRIPTION
**Pull request by gertjaap**
This pull request forks the mainnet at block 1500000 to Verthash. It will reset the difficulty and lower the powLimit with a factor 128 compared to the Lyra2REv3 hardfork. This is done based on the results of running Verthash on the same GPUs and comparing hashrates:

image

### CODE REVIEW

**Comment by metalicjames** [comment]
```
../../src/crypto/verthash.cpp: In static member function ‘static void Verthash::Hash(const char*, char*)’:
../../src/crypto/verthash.cpp:128:68: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
                 fread(&value, sizeof(uint32_t), 1, VerthashDatFile);
                                                                    ^
```

Do you think we should fix this warning? Silently failing an fread here could be bad.

**Comment by metalicjames** [comment]
This looks good to me other than the build warning in verthash.cpp.

**Comment by gertjaap** [comment]
> ```
> ../../src/crypto/verthash.cpp: In static member function ‘static void Verthash::Hash(const char*, char*)’:
> ../../src/crypto/verthash.cpp:128:68: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’, declared with attribute warn_unused_result [-Wunused-result]
>                  fread(&value, sizeof(uint32_t), 1, VerthashDatFile);
>                                                                     ^
> ```
> 
> Do you think we should fix this warning? Silently failing an fread here could be bad.

Definitely should fix. I've added an assertion on the result (`read_len == 1`)

File: src/chainparams.cpp [hunk 1]
```diff
@@ -75,7 +75,7 @@ class CMainParams : public CChainParams {
         
         // TODO Determine correct value based on average miner performance / expected nethash
```
Comment (metalicjames) [inline src/chainparams.cpp]: Remove this todo?

File: src/crypto/verthash.cpp [hunk 2]
```diff
@@ -120,12 +120,14 @@ void Verthash::Hash(const char* input, char* output)
     const uint32_t mdiv = ((datfile_sz - HASH_OUT_SIZE)/BYTE_ALIGNMENT) + 1;
     
     if(!datFileInRam) {
+        size_t read_len = -1;
```
Comment (metalicjames) [inline src/crypto/verthash.cpp]: Any reason not to scope this variable to the same line as the fread?

File: src/pow.cpp [hunk 3]
```diff
@@ -35,12 +35,9 @@ unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHead
...
         } else if(nHeight >= 1080000 && nHeight < 1080010) { // Force difficulty for 10 blocks - Lyra2REv3 hardfork
             return 0x1b0ffff0;
-        } 
-        /* 
-            else if(nHeight >= VERTHASH_FORKBLOCK_MAINNET && nHeight < VERTHASH_FORKBLOCK_MAINNET+10) { // Force difficulty for 10 blocks - Verthash hardfork
-                return 0x1b0ffff0;
-            }
-        */
+        } else if(nHeight >= VERTHASH_FORKBLOCK_MAINNET && nHeight < VERTHASH_FORKBLOCK_MAINNET+10) { // Force difficulty for 10 blocks - Verthash hardfork
+            return 0x1c07fff8;
```
Comment (metalicjames) [inline src/pow.cpp:39]: How is this number calculated?

File: src/pow.cpp [hunk 4]
```diff
@@ -163,12 +160,11 @@ unsigned int KimotoGravityWell(const CBlockIndex* pindexLast,
     }
 
     if(Params().NetworkIDString() == CBaseChainParams::MAIN) {
-        // Code to activate mainnet fork
-        /*if((uint64_t)BlockLastSolved->nHeight+1 >= VERTHASH_FORKBLOCK_MAINNET) {
+        if((uint64_t)BlockLastSolved->nHeight+1 >= VERTHASH_FORKBLOCK_MAINNET) {
```
Comment (metalicjames) [inline src/pow.cpp]: This would be better as `static_cast`.

**Review by metalicjames — CHANGES_REQUESTED** [review]
Sorry, I went back over it and have a few more comments.

**Comment by metalicjames** [comment]
Can you re-target this PR to master? I replaced the ancient branch there to be equivalent to 0.16.0.

File: src/crypto/verthash.cpp (hunk 2 above)
Comment (gertjaap) [reply src/crypto/verthash.cpp]: I was (wrongly) assuming this causes overhead - but it looks like compilers optimize this away. Changed

File: src/chainparams.cpp (hunk 1 above)
Comment (gertjaap) [reply src/chainparams.cpp]: Yep

File: src/pow.cpp (hunk 3 above)
Comment (gertjaap) [reply src/pow.cpp:39]: See the opening comments of the PR for the factor 128 explanation.

I calculated the factor 128 against the previous hardfork difficulty like this:

* Convert compact bits (0x1b0ffff0) to bignum target 
* Multiply resulting bignum by 128
* Convert back to compact bits

https://play.golang.org/p/TVmR62Gr071

File: src/pow.cpp (hunk 4 above)
Comment (gertjaap) [reply src/pow.cpp]: Changed

**Review by metalicjames — APPROVED** [review]

**Pull request merged**